File: 01-ll-designs/consistent_hashing/solutions/python/consistent_hash.py
import hashlib
import bisect
from typing import Dict, List, Optional, Set
from collections import defaultdict
import threading


class ConsistentHash:
    """
    Consistent Hashing implementation with virtual nodes support.
    
    This implementation uses MD5 hashing to distribute keys and nodes
    across a hash ring, with virtual nodes to improve load distribution.
    """
    
    def __init__(self, virtual_nodes: int = 3):
        """
        Initialize the consistent hash ring.
        
        Args:
            virtual_nodes: Number of virtual nodes per physical node
        """
        self.virtual_nodes = virtual_nodes
        self.ring: Dict[int, str] = {}  # hash -> node_id
        self.sorted_hashes: List[int] = []  # sorted hash values
        self.nodes: Set[str] = set()  # active nodes
        self.lock = threading.RLock()  # thread safety
    
    def _hash(self, key: str) -> int:
        """Generate hash value for a key using MD5."""
        return int(hashlib.md5(key.encode('utf-8')).hexdigest(), 16)
    
    def add_node(self, node_id: str) -> None:
        """
        Add a node to the hash ring.
        
        Args:
            node_id: Unique identifier for the node
        """
        with self.lock:
            if node_id in self.nodes:
                return
            
            self.nodes.add(node_id)
            
            # Add virtual nodes to the ring
            for i in range(self.virtual_nodes):
                virtual_key = f"{node_id}:{i}"
                hash_value = self._hash(virtual_key)
                
                self.ring[hash_value] = node_id
                bisect.insort(self.sorted_hashes, hash_value)
    
    def remove_node(self, node_id: str) -> None:
        """
        Remove a node from the hash ring.
        
        Args:
            node_id: Identifier of the node to remove
        """
        with self.lock:
            if node_id not in self.nodes:
                return
            
            self.nodes.remove(node_id)
            
            # Remove virtual nodes from the ring
            for i in range(self.virtual_nodes):
                virtual_key = f"{node_id}:{i}"
                hash_value = self._hash(virtual_key)
                
                if hash_value in self.ring:
                    del self.ring[hash_value]
                    self.sorted_hashes.remove(hash_value)
    
    def get_node(self, key: str) -> Optional[str]:
        """
        Get the node responsible for a given key.
        
        Args:
            key: The key to look up
            
        Returns:
            Node ID that should handle the key, or None if no nodes available
        """
        with self.lock:
            if not self.sorted_hashes:
                return None
            
            hash_value = self._hash(key)
            
            # Find the first node clockwise from the key's hash
            idx = bisect.bisect_right(self.sorted_hashes, hash_value)
            
            # Wrap around to the beginning if we're past the end
            if idx == len(self.sorted_hashes):
                idx = 0
            
            return self.ring[self.sorted_hashes[idx]]
    
    def get_nodes(self) -> List[str]:
        """Get all active nodes in the system."""
        with self.lock:
            return list(self.nodes)
    
    def get_load_distribution(self, keys: List[str]) -> Dict[str, int]:
        """
        Analyze load distribution for a set of keys.
        
        Args:
            keys: List of keys to analyze
            
        Returns:
            Dictionary mapping node_id to number of keys assigned
        """
        distribution = defaultdict(int)
        
        for key in keys:
            node = self.get_node(key)
            if node:
                distribution[node] += 1
        
        return dict(distribution)
    
    def get_ring_info(self) -> Dict:
        """Get information about the current ring state."""
        with self.lock:
            return {
                'total_nodes': len(self.nodes),
                'total_virtual_nodes': len(self.sorted_hashes),
                'virtual_nodes_per_node': self.virtual_nodes,
                'nodes': list(self.nodes)
            }


def demonstrate_consistent_hashing():
    """Demonstrate consistent hashing functionality."""
    print("=== Consistent Hashing Demo ===\n")
    
    # Create hash ring
    ch = ConsistentHash(virtual_nodes=3)
    
    # Add initial nodes
    nodes = ["server1", "server2", "server3"]
    for node in nodes:
        ch.add_node(node)
    
    print(f"Added nodes: {nodes}")
    print(f"Ring info: {ch.get_ring_info()}\n")
    
    # Test key distribution
    test_keys = [f"user:{i}" for i in range(1, 11)]
    
    print("Initial key distribution:")
    distribution = ch.get_load_distribution(test_keys)
    for node, count in distribution.items():
        print(f"  {node}: {count} keys")
    
    # Show specific key mappings
    print("\nKey mappings:")
    for key in test_keys[:5]:
        node = ch.get_node(key)
        print(f"  {key} -> {node}")
    
    # Remove a node and show redistribution
    print(f"\nRemoving 'server2'...")
    ch.remove_node("server2")
    
    print("New key distribution:")
    new_distribution = ch.get_load_distribution(test_keys)
    for node, count in new_distribution.items():
        print(f"  {node}: {count} keys")
    
    # Calculate redistribution
    redistributed = 0
    for key in test_keys:
        old_node = None
        for node, count in distribution.items():
            if node == "server2":
                continue
        new_node = ch.get_node(key)
        # This is a simplified check - in practice you'd track the changes
    
    print(f"\nRing info after removal: {ch.get_ring_info()}")
    
    # Add a new node
    print(f"\nAdding 'server4'...")
    ch.add_node("server4")
    
    final_distribution = ch.get_load_distribution(test_keys)
    print("Final key distribution:")
    for node, count in final_distribution.items():
        print(f"  {node}: {count} keys")


if __name__ == "__main__":
    demonstrate_consistent_hashing()

File: 01-ll-designs/consistent_hashing/solutions/cpp/consistent_hash.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <sstream>
#include <iomanip>

#ifndef USE_STD_HASH
#include <openssl/md5.h>
#endif

/**
 * Consistent Hashing implementation with virtual nodes support.
 * 
 * This implementation uses MD5 hashing (or std::hash as fallback) to distribute 
 * keys and nodes across a hash ring, with virtual nodes to improve load distribution.
 */
class ConsistentHash {
private:
    int virtualNodes;
    std::map<uint64_t, std::string> ring;  // hash -> node_id
    std::set<std::string> nodes;           // active nodes
    
    /**
     * Generate hash value for a key.
     * Uses MD5 if available, otherwise falls back to std::hash.
     */
    uint64_t hash(const std::string& key) const {
#ifndef USE_STD_HASH
        // Use MD5 for better distribution
        unsigned char digest[MD5_DIGEST_LENGTH];
        MD5(reinterpret_cast<const unsigned char*>(key.c_str()), key.length(), digest);
        
        uint64_t hashValue = 0;
        for (int i = 0; i < 8; i++) {
            hashValue = (hashValue << 8) | digest[i];
        }
        return hashValue;
#else
        // Fallback to std::hash
        return std::hash<std::string>{}(key);
#endif
    }
    
public:
    /**
     * Initialize the consistent hash ring.
     * 
     * @param virtualNodes Number of virtual nodes per physical node
     */
    explicit ConsistentHash(int virtualNodes = 3) : virtualNodes(virtualNodes) {}
    
    /**
     * Add a node to the hash ring.
     * 
     * @param nodeId Unique identifier for the node
     */
    void addNode(const std::string& nodeId) {
        if (nodes.find(nodeId) != nodes.end()) {
            return;  // Node already exists
        }
        
        nodes.insert(nodeId);
        
        // Add virtual nodes to the ring
        for (int i = 0; i < virtualNodes; i++) {
            std::string virtualKey = nodeId + ":" + std::to_string(i);
            uint64_t hashValue = hash(virtualKey);
            ring[hashValue] = nodeId;
        }
    }
    
    /**
     * Remove a node from the hash ring.
     * 
     * @param nodeId Identifier of the node to remove
     */
    void removeNode(const std::string& nodeId) {
        if (nodes.find(nodeId) == nodes.end()) {
            return;  // Node doesn't exist
        }
        
        nodes.erase(nodeId);
        
        // Remove virtual nodes from the ring
        for (int i = 0; i < virtualNodes; i++) {
            std::string virtualKey = nodeId + ":" + std::to_string(i);
            uint64_t hashValue = hash(virtualKey);
            ring.erase(hashValue);
        }
    }
    
    /**
     * Get the node responsible for a given key.
     * 
     * @param key The key to look up
     * @return Node ID that should handle the key, or empty string if no nodes available
     */
    std::string getNode(const std::string& key) const {
        if (ring.empty()) {
            return "";
        }
        
        uint64_t hashValue = hash(key);
        
        // Find the first node clockwise from the key's hash
        auto it = ring.lower_bound(hashValue);
        
        // Wrap around to the beginning if we're past the end
        if (it == ring.end()) {
            it = ring.begin();
        }
        
        return it->second;
    }
    
    /**
     * Get all active nodes in the system.
     * 
     * @return Vector of active node IDs
     */
    std::vector<std::string> getNodes() const {
        return std::vector<std::string>(nodes.begin(), nodes.end());
    }
    
    /**
     * Analyze load distribution for a set of keys.
     * 
     * @param keys Vector of keys to analyze
     * @return Map of node_id to number of keys assigned
     */
    std::unordered_map<std::string, int> getLoadDistribution(const std::vector<std::string>& keys) const {
        std::unordered_map<std::string, int> distribution;
        
        for (const auto& key : keys) {
            std::string node = getNode(key);
            if (!node.empty()) {
                distribution[node]++;
            }
        }
        
        return distribution;
    }
    
    /**
     * Get information about the current ring state.
     * 
     * @return String containing ring statistics
     */
    std::string getRingInfo() const {
        std::ostringstream oss;
        oss << "Total nodes: " << nodes.size()
            << ", Total virtual nodes: " << ring.size()
            << ", Virtual nodes per node: " << virtualNodes;
        return oss.str();
    }
    
    /**
     * Print the current state of the hash ring (for debugging).
     */
    void printRing() const {
        std::cout << "Hash Ring State:\n";
        for (const auto& entry : ring) {
            std::cout << "  Hash: " << std::hex << entry.first 
                      << " -> Node: " << entry.second << std::dec << "\n";
        }
    }
};

/**
 * Demonstrate consistent hashing functionality.
 */
void demonstrateConsistentHashing() {
    std::cout << "=== Consistent Hashing Demo ===\n\n";
    
    // Create hash ring
    ConsistentHash ch(3);
    
    // Add initial nodes
    std::vector<std::string> nodes = {"server1", "server2", "server3"};
    for (const auto& node : nodes) {
        ch.addNode(node);
    }
    
    std::cout << "Added nodes: ";
    for (const auto& node : nodes) {
        std::cout << node << " ";
    }
    std::cout << "\n";
    std::cout << "Ring info: " << ch.getRingInfo() << "\n\n";
    
    // Test key distribution
    std::vector<std::string> testKeys;
    for (int i = 1; i <= 10; i++) {
        testKeys.push_back("user:" + std::to_string(i));
    }
    
    std::cout << "Initial key distribution:\n";
    auto distribution = ch.getLoadDistribution(testKeys);
    for (const auto& entry : distribution) {
        std::cout << "  " << entry.first << ": " << entry.second << " keys\n";
    }
    
    // Show specific key mappings
    std::cout << "\nKey mappings:\n";
    for (int i = 0; i < std::min(5, static_cast<int>(testKeys.size())); i++) {
        const auto& key = testKeys[i];
        std::string node = ch.getNode(key);
        std::cout << "  " << key << " -> " << node << "\n";
    }
    
    // Remove a node and show redistribution
    std::cout << "\nRemoving 'server2'...\n";
    ch.removeNode("server2");
    
    std::cout << "New key distribution:\n";
    auto newDistribution = ch.getLoadDistribution(testKeys);
    for (const auto& entry : newDistribution) {
        std::cout << "  " << entry.first << ": " << entry.second << " keys\n";
    }
    
    std::cout << "\nRing info after removal: " << ch.getRingInfo() << "\n";
    
    // Add a new node
    std::cout << "\nAdding 'server4'...\n";
    ch.addNode("server4");
    
    auto finalDistribution = ch.getLoadDistribution(testKeys);
    std::cout << "Final key distribution:\n";
    for (const auto& entry : finalDistribution) {
        std::cout << "  " << entry.first << ": " << entry.second << " keys\n";
    }
    
    // Demonstrate load balancing with more keys
    std::cout << "\n=== Load Balancing Test ===\n";
    std::vector<std::string> manyKeys;
    for (int i = 1; i <= 1000; i++) {
        manyKeys.push_back("key:" + std::to_string(i));
    }
    
    auto loadTest = ch.getLoadDistribution(manyKeys);
    std::cout << "Distribution of 1000 keys:\n";
    for (const auto& entry : loadTest) {
        double percentage = (entry.second * 100.0) / manyKeys.size();
        std::cout << "  " << entry.first << ": " << entry.second 
                  << " keys (" << std::fixed << std::setprecision(1) 
                  << percentage << "%)\n";
    }
}

int main() {
    demonstrateConsistentHashing();
    return 0;
}